articles/setting-up-own-login-vs-login-via-external-services/index.md

```diff
@@ -0,0 +1,68 @@
+
+### Basic Setup for Login-Signup Flow
+ Login-Signup is the first line of security for any service. If this is breached, a malicious user can access data that is otherwise not allowed. The login interface should also verify for user access level during the process- vendors, delivery boy, buyer, etc. in case of e-commerce or editor, author, a viewer in case of a blog, and similarly for other services. At the bare minimum, the authentication should be on username/email with the password, which can be secured even further with multi-factor authentication. To ensure privacy and protection of the users, passwords must be hashed at all stages(even before sendingit to the server via API).
+
+### Some pointers to keep in mind while hashing are:
+* Use hash function, not encryption, no matter how secure the encryption is. There is no feature or facility required for login and sign-up flow that is not provided by hashing methodology which encryption provides.
+* Use powerful hashing methods with salt(random salt preferably; having the same salt for all passwords makes salting go in vain). Hashing algorithms like MD5 are too weak for present-day computational power, hence comparatively insecure.
+* Additionally, older standards of hashing have multiple known collision pairs. A collision is an event where two distinct inputs give the same hash as output.
+* If you are using API and passing hashed password in request(hashed passwords always better than plain-text), then do not directly store this has in server. If you directly store the hash, then your entire database if compromised in multiple ways.  
+  * Any internal employee with access to the database can copy hash from the database and use API to login.
+  * If the database gets breached by a hacker, they no longer need to get a plain-text password to login as API can do it directly
+
+ Hence, you should hash passwords at each level of transmission, multiple times if needed.
+
+### Additional Ways to Secure
+* Go Slow. In this age of high-speed internet, how can going slow be beneficial?
+  * A wrong password entry should have a cool-off time before next attempt. This can be done from first attempt itself(a few hundred milliseconds) or after few attempts(maybe three of five). The cool-down time would be decided after coming to a common ground between User Experience and Security standpoints. This cool-down should be done on the server end and not client end as client-side scripts are easier to manipulate. Additionally, is done client-side, a script can have multiple instances of a client running parallelly, which will bypass client-side protection.
+  * IP Rate Limiting. This should be used to protect bogus sign-ups from the same IP. Additionally, multiple users trying to log in from the same IP would also be an indication of compromised accounts.
```

Comment:
In corporate environments, educational institutions, countries with Internet filtering, users using VPNs, and even multiple people sharing Wifi in their home, it is perfectly normal for many users to have the same IP address. Rate limiting is still useful but the thresholds need to be higher to allow for these common legitimate scenarios.